server/src/main/java/org/elasticsearch/env/NodeMetadata.java

```diff
@@ -92,14 +92,16 @@ public Version previousNodeVersion() {
     }
 
     public NodeMetadata upgradeToCurrentVersion() {
-        if (nodeVersion.equals(Version.V_EMPTY)) {
```

Comment:
Could we `assert nodeVersion.equals(Version.V_EMPTY) == false` to ensure that we're right to drop this branch?

Reply:
We have a test that tests this explicitly - testUpgradesMissingVersion, we throw the same version exception as if the version was 7.nonlast, since V_EMPTY resolves to 0. 

I can swap the test with the assert, but I think the test is a stronger validation that we'll fail to start with <= 6.latest metadata, where version is empty as per the previous assert.

I think I'll introduce an assert like the one we had before, ensuring that Version can only be empty if we are older than 7.0 + 1.